vpr/src/analytical_place/analytical_solver.cpp

```diff
@@ -0,0 +1,264 @@
+                               unsigned iteration) {
+    // Anchor weights grow exponentially with iteration.
+    double coeff_pseudo_anchor = 0.01 * std::exp((double)iteration/5);
+    for (size_t row_id_idx = 0; row_id_idx < num_moveable_blocks; row_id_idx++) {
+        APRowId row_id = APRowId(row_id_idx);
+        APBlockId blk_id = row_id_to_blk_id[row_id];
+        double pseudo_w = coeff_pseudo_anchor;
+        A_sparse_diff.coeffRef(row_id_idx, row_id_idx) += pseudo_w;
+        b_x_diff(row_id_idx) += pseudo_w * p_placement.block_x_locs[blk_id];
+        b_y_diff(row_id_idx) += pseudo_w * p_placement.block_y_locs[blk_id];
+    }
+}
+
+void QPHybridSolver::solve(unsigned iteration, PartialPlacement &p_placement) {
+    // Create a temporary linear system which will contain the original linear
+    // system which may be updated to include the anchor points.
+    Eigen::SparseMatrix<double> A_sparse_diff = Eigen::SparseMatrix<double>(A_sparse);
+    Eigen::VectorXd b_x_diff = Eigen::VectorXd(b_x);
+    Eigen::VectorXd b_y_diff = Eigen::VectorXd(b_y);
+    // In the first iteration, the orginal linear system is used.
+    // In any other iteration, use the moveable APBlocks current placement as
+    //                         anchor-points (fixed block positions).
+    if (iteration != 0) {
+        update_linear_system_with_anchors(A_sparse_diff, b_x_diff, b_y_diff,
+                                          p_placement, num_moveable_blocks_,
+                                          row_id_to_blk_id_, iteration);
+    }
+    // Verify that the constant vectors are valid.
+    VTR_ASSERT_DEBUG(!b_x_diff.hasNaN() && "b_x has NaN!");
+    VTR_ASSERT_DEBUG(!b_y_diff.hasNaN() && "b_y has NaN!");
+
+    // Set up the ConjugateGradient Solver using the coefficient matrix.
+    // TODO: can change cg.tolerance to increase performance when needed
+    //  - This tolerance may need to be a function of the number of nets.
+    //  - Instead of normalizing the fixed blocks, the tolerance can be scaled
+    //    by the size of the device.
+    Eigen::ConjugateGradient<Eigen::SparseMatrix<double>, Eigen::Lower|Eigen::Upper> cg;
+    cg.compute(A_sparse_diff);
+    VTR_ASSERT(cg.info() == Eigen::Success && "Conjugate Gradient failed at compute!");
+    // Use the solver to solve for x and y using the constant vectors
+    // TODO: Use solve with guess to make this faster. Use the previous placement
+    //       as a guess.
+    Eigen::VectorXd x = cg.solve(b_x_diff);
+    VTR_ASSERT(cg.info() == Eigen::Success && "Conjugate Gradient failed at solving b_x!");
+    Eigen::VectorXd y = cg.solve(b_y_diff);
+    VTR_ASSERT(cg.info() == Eigen::Success && "Conjugate Gradient failed at solving b_y!");
+
+    // Write the results back into the partial placement object.
+    for (size_t row_id_idx = 0; row_id_idx < num_moveable_blocks_; row_id_idx++) {
+        APRowId row_id = APRowId(row_id_idx);
```

Comment:
This code assumes that the first rows are all for movable blocks. I think you commented that elsewhere so it's not the end of the world, but you could rewrite this code to go through all blocks and get their mapping to rows, so it wouldn't have any such assumption.

Reply:
This was actually done for performance. I think iterating linearly from 0 to num_moveable_blocks would be more efficient than iterating over all blocks and skipping the fixed blocks. I left a comment explaining that the first rows must represent the moveable blocks and added debug asserts to enforce this.